dsdl_compiler/libcanard_dsdl_compiler/code_type_template.tmpl

```diff
@@ -0,0 +1,228 @@
+    /*  - Add array length, last item, but bitlen < 8. */
+    canardEncodeScalar(msg_buf, offset, ${f.array_max_size_bit_len}, (void*)&source->${'%s' % ((f.name + '_len'))});
+    offset += ${f.array_max_size_bit_len};
+                    %else
+    if (! root_item) {
+        /* - Add array length */
+        canardEncodeScalar(msg_buf, offset, ${f.array_max_size_bit_len}, (void*)&source->${'%s' % ((f.name + '_len'))});
+        offset += ${f.array_max_size_bit_len};
+    }
+                    %endif
+                %else
+    /* - Add array length */
+    canardEncodeScalar(msg_buf, offset, ${f.array_max_size_bit_len}, (void*)&source->${'%s' % ((f.name + '_len'))});
+    offset += ${f.array_max_size_bit_len};
+                %endif
+    /* - Add array items */
+    for (c = 0; c < source->${'%s' % ((f.name + '_len'))}; c++) {
+                %if f.cpp_type_category == t.CATEGORY_COMPOUND:
+        offset += ${f.cpp_type}Encode((void*)&source->${f.name}[c], msg_buf, offset, 0);
+                %else
+        canardEncodeScalar(msg_buf, offset, ${f.bitlen}, (void*)(source->${f.name} + c)); // ${f.max_size}
+        offset += ${f.bitlen};
+                %endif
+    }
+            %else
+    /* Static array (${f.name}) */
+    for (c = 0; c < ${f.array_size}; c++) {
+        canardEncodeScalar(msg_buf, offset, ${f.bitlen}, (void*)(source->${f.name} + c)); // ${f.max_size}
+        offset += ${f.bitlen};
+    }
+            %endif
+
+        %elif f.type_category == t.CATEGORY_VOID:
+
+    /* Void${f.bitlen} */
+    offset += ${f.bitlen};
+        %elif f.type_category == t.CATEGORY_COMPOUND:
+
+    /* Compound */
+    offset = ${f.cpp_type}Encode((void*)&source->${f.name}, msg_buf, offset, 0);
+        %else
+            %if f.saturate
+    source->${f.name} = SATURATE(source->${f.name}, ${f.max_size})
+            %endif
+    canardEncodeScalar(msg_buf, offset, ${f.bitlen}, (void*)&source->${f.name}); // ${f.max_size}
+    offset += ${f.bitlen};
+
+        %endif
+     %if union:
+    }
+     %endif
+    % endfor
+    if (root_item) {
+        return (offset + 7 ) / 8;
+    }
+
+    return offset;
+}
+
+/**
+  * @brief ${type_name}Decode
+  * @param transfer: Pointer to CanardRxTransfer transfer
+  * @param payload_len: Payload message length
+  * @param dest: Pointer to destination struct
+  * @param dyn_arr_buf: NULL or Pointer to memory storage to be used for dynamic arrays
+  *                     ${type_name} dyn memory will point to dyn_arr_buf memory.
+  *                     NULL will ignore dynamic arrays decoding.
+  * @param offset: Call with 0, bit offset to msg storage
+  * @retval
+  */
+uint32_t ${type_name}Decode(const CanardRxTransfer* transfer, uint16_t payload_len, ${type_name} *dest, uint8_t **dyn_arr_buf, uint32_t offset) {
+    %if has_array
+    uint32_t c = 0;
+    %endif
+
+    %if union:
+    /* Get Union Tag */
+    canardDecodeScalar(transfer, offset, ${union}, false, (void*)&dest->union_tag); // ${union}
```

Comment:
Here and in several other places, the return value is not checked. It should be. If the returned value is non-positive or does not equal the number of bits requested, the operation should be aborted (MISRA 0-1-7).